### PR TITLE
Add more fine-grained prefetch() directive (Issue #3735)

### PR DESCRIPTION
**Pull request by steven-johnson**
This adds a new scheduling directive, `prefetch_at()`, that allows you to specify prefetching a value from a different loop variable (See issue #3735 for details). Note that `prefetch_at(f, b, b)` is exactly identical to `prefetch(f, b)` in function.

Opening as draft for now:

~~Defining it as an overload on `prefetch()` isn't possible due to ambiguous overloads that would be created, hence the new name; since the existing `prefetch()` directive is rarely used, do we just change its argument list rather than adding a wrapper?~~

~~Doc comments needed~~

~~Python API wrappers needed~~

- Changing the parameters to __builtin_prefetch are a *big* win, at least on ARM experiments. Do we need to expose these knobs to user code somehow? (Do we know why the previous values where specified?)

- No testing on Hexagon has been done.

- Initial testing on ARM indicates that adding prefetch can add a surprising amount of apparently-unnecessary stack spill from the generated code; this needs to be investigated.

### CODE REVIEW

File: src/runtime/prefetch.cpp [hunk 1]
```diff
@@ -6,7 +6,9 @@ extern "C" {
 // parameter breaks a lot of optimizations, but needs to be WEAK
 // so that Codegen_LLVM can find an instance of the Function to insert.
 WEAK_INLINE int _halide_prefetch(const void *ptr) {
-    __builtin_prefetch(ptr, 1, 3);
+    constexpr int rw = 0;        // 1 = write, 0 = read
+    constexpr int locality = 0;  // 0 = no temporal locality, 3 = high temporal locality
```
Comment (dsharletg) [inline src/runtime/prefetch.cpp:10]: I wonder if we need to make these options parameters to the prefetch directive somehow. Experimenting with them does produce different results. However, I suspect these current values (read, no temporal locality) is probably the most useful/common case.

**Comment by steven-johnson** [comment]
The apparently-unnecessary stack spill has been reported to LLVM as https://bugs.llvm.org/show_bug.cgi?id=51172

**Comment by steven-johnson** [comment]
Marked as read for review because it's basically doing what we want -- the issue with LLVM is one that (hopefully) will improve code generation further, but it isn't a blocker, strictly speaking.

File: src/Func.h [hunk 2]
```diff
@@ -1990,13 +2003,101 @@ class Func {
...
-        return prefetch(image.parameter(), var, offset, strategy);
+        return prefetch_at<T>(image, var, var, std::move(offset), strategy);
+    }
+    // @}
+
+    /** prefetch_at() is a more fine-grained version of prefetch(), which allows
+     * specification of different vars for the location of the prefetch() instruction
+     * vs. the location that is being prefetched:
+     *
+     * - the first var specified, 'loop', indicates the loop in which the prefetch will be placed
```
Comment (dsharletg) [inline src/Func.h:2024]: I think there must be better names for these, at least `loop`. Both options are loops... I think the first var should be something related to `at`, given the similarity to `compute_at` and `store_at`. `prefetch_at`, and `prefetch_from` for the other variable? Or maybe just `at` and `from` given the context.

File: src/Prefetch.cpp [hunk 3]
```diff
@@ -154,8 +155,10 @@ class InjectPrefetch : public IRMutator {
         } else if (op->bounds.empty()) {
             // Remove the Prefetch IR since it is prefetching an empty region
             user_warning << "Removing prefetch of " << p.name
-                         << " within loop nest of " << p.var << " (offset: "
-                         << p.offset << ") since it is not used at all.\n";
+                         << " within loop nest of " << p.loop_var
+                         << " at location " << p.fetch_var
```
Comment (dsharletg) [inline src/Prefetch.cpp:159]: at -> from?

File: src/runtime/qurt_hvx.cpp [hunk 4]
```diff
@@ -57,7 +57,7 @@ WEAK_INLINE int _halide_prefetch_2d(const void *ptr, int width_bytes, int height
     return 0;
 }
 
-WEAK_INLINE int _halide_prefetch(const void *ptr, int size) {
+WEAK_INLINE int _halide_prefetch_1D(const void *ptr, int size) {
```
Comment (dsharletg) [inline src/runtime/qurt_hvx.cpp:60]: 1D -> 1d (the call site is 1d, and matches 2d convention)

File: src/Func.h (hunk 2 above)
Comment (steven-johnson) [reply src/Func.h]: Used `at` and `from`, PTAL

File: src/Prefetch.cpp (hunk 3 above)
Comment (steven-johnson) [reply src/Prefetch.cpp]: Done

File: src/runtime/qurt_hvx.cpp (hunk 4 above)
Comment (steven-johnson) [reply src/runtime/qurt_hvx.cpp]: Done

File: src/Func.cpp [hunk 5]
```diff
@@ -1840,14 +1840,14 @@ Stage &Stage::hexagon(const VarOrRVar &x) {
     return *this;
 }
 
-Stage &Stage::prefetch(const Func &f, const VarOrRVar &var, Expr offset, PrefetchBoundStrategy strategy) {
-    PrefetchDirective prefetch = {f.name(), var.name(), std::move(offset), strategy, Parameter()};
+Stage &Stage::prefetch_at(const Func &f, const VarOrRVar &loop, const VarOrRVar &from, Expr offset, PrefetchBoundStrategy strategy) {
```
Comment (dsharletg) [inline src/Func.cpp:1843]: Rename these too, for consistency with the header?

File: src/Func.h [hunk 6]
```diff
@@ -1990,13 +2003,101 @@ class Func {
...
+     * The following schedule:
+     \code
+     f.compute_root();
+     g.prefetch_at(f, x, x, 2, PrefetchBoundStrategy::NonFaulting);
+     h.prefetch_at(f, x, y, 2, PrefetchBoundStrategy::NonFaulting);
+     \endcode
+     *
+     * will inject prefetch call at the innermost loop of 'g' and 'h' and generate
+     * the following loop nest:
+     * for y = ...
```
Comment (dsharletg) [inline src/Func.h:2049]: \code...\endcode

File: src/Func.h [hunk 7]
```diff
@@ -1990,13 +2003,101 @@ class Func {
...
+     \endcode
+     *
+     * The following schedule:
+     \code
+     f.compute_root();
+     g.prefetch_at(f, y, w, 2, PrefetchBoundStrategy::NonFaulting);
+     \endcode
+     *
+     * will produce code that prefetches a tile of data:
+     * for w = ...
```
Comment (dsharletg) [inline src/Func.h:2076]: \code...\endcode

File: src/Func.h [hunk 8]
```diff
@@ -1990,13 +2003,101 @@ class Func {
...
+     *   for z = ...
+     *     for y = ...
+     *       for x0 = ...
+     *          prefetch(&f[x0, y, z, w + 2], 1, 16);
+     *       for x = ...
+     *         g(x, y, z, w) = 2 * f(x, y, z, w)
+     *
+     * Note that calling prefetch_at() with the same var for both 'at' and 'from'
+     * is exactly equivalent to calling prefetch() with that var.
```
Comment (dsharletg) [inline src/Func.h:2090]: Nit: I'd remove 'exactly' here.

**Review by dsharletg — APPROVED** [review]

**Comment by steven-johnson** [comment]
Per discussion at group meeting:
- Existing version of prefetch() had `offset` changed from Expr to int (as a temporary measure to avoid ambiguous call resolution issues), then marked as deprecated
- The new, two-var call is also named prefetch(), but is not deprecated
- All existing calls within the Halide codebase that reference the deprecated call have been updated to the non-deprecated call

**Comment by steven-johnson** [comment]
Do we want to go ahead and land this (in advance of hoped-for LLVM improvements?)

**Comment by dsharletg** [comment]
I think we should, I think the change is working as designed in Halide. And, despite the LLVM codegen issues, this was still a big improvement in a few cases I tried.

**Pull request merged**